File: preprocessing_matlab/matlabScripts/fromArcGIS.m
% FromArcGIS.M
% Programm zum Einlesen der Daten aus ArcGIS
% This class is part of the catflow wizard.
% See the help-file catflow.chm for more information

% Copyright (C) 2005  Dominik Reusser (domi_reusser@gmx.net)

% This program is free software; you can redistribute it and/or
% modify it under the terms of the GNU General Public License
% as published by the Free Software Foundation; either version 2
% of the License, or (at your option) any later version.
 
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.

% You should have received a copy of the GNU General Public License
% along with this program; if not, write to the Free Software
% Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
% 
% Input:  Wird abgefragt. ASCII-File aus GIS erzeugt
%

%clear 
global htyp

    
    [filename, pathname] = uigetfile('*.txt', 'Pick the file produced by ArcGIS');
    if isequal(filename,0)
       disp('User selected Cancel')
       exit
   end
    fid = fopen(fullfile(pathname, filename));
  %Dataformat from ArcGis preprocessing:
  %First line: numberOfSlopes (n), maxNumberOfSlopePoints, x_bez, y_bez, z_bez, number of Xsi, number of Eta
    slopeCount = fscanf(fid, '%i',[1 1]) 
    slopePointCount = fscanf(fid, '%i',[1,1])
    re_bez = fscanf(fid, '%g',[1,1])
    ho_bez = fscanf(fid, '%g',[1,1])
    z_bez = fscanf(fid, '%g',[1,1])
    xsiCount = fscanf(fid, '%i',[1,1])
    etaCount = fscanf(fid, '%i',[1,1])
    %line 2 SlopeID
    numh = fscanf(fid, '%i',[1,slopeCount])
  %line 3 SlopeType (Constant Thickness or 'Cake shape')
    htyp = fscanf(fid, '%i',[1,slopeCount])
  %line 4 Area
    flaech = fscanf(fid, '%g',[1,slopeCount])
    flaech = flaech'
  %line 5 Number of points
    punkteh = fscanf(fid, '%i',[1,slopeCount])
  %line 6 Thickness of slope
    dyyy = fscanf(fid, '%g',[1,slopeCount])
    
  %line 7 Eta Values
    eta = fscanf(fid, '%g',[1,etaCount])
  %line 8 Xsi Values
    xsi = fscanf(fid, '%g',[1,xsiCount])
    xsi = xsi'
    if(not(issorted(eta)))
        error('eta is not sorted ascending')
    end
    if(not(issorted(xsi)))
        error('xsi is not sorted ascending')
    end
  %line 9...8+n x of slopepoints
    xh = fscanf(fid, '%g',[slopePointCount,slopeCount])
  %line 9+n...8+2n y of slopepoints
    yh = fscanf(fid, '%g',[slopePointCount,slopeCount])
    
  %line 9+2n...8+3n z of slopepoints
    zh = fscanf(fid, '%g',[slopePointCount,slopeCount])
  %line 9+3n...8+4n width of slopepoints
    bh = fscanf(fid, '%g',[slopePointCount,slopeCount])
    
     fclose(fid)

File: preprocessing_RCatflow/R/fromArcGIS.R
fromArcGIS <-
function(filename)
{  
  erg <- list()

  dat <- readLines(filename)
   
  #Dataformat from ArcGis preprocessing:
  #First line: numberOfSlopes (n), maxNumberOfSlopePoints, x_bez, y_bez, z_bez, number of Xsi, number of Eta
    firstLine <- as.numeric(na.exclude(as.numeric(strsplit(dat[1]," ")[[1]])))
      slopeCount = firstLine[1]
      slopePointCount = firstLine[2]
      re_bez = firstLine[3]
      ho_bez = firstLine[4]
      z_bez = firstLine[5]
      xsiCount = firstLine[6]
      etaCount = firstLine[7]
  
  if(slopeCount == 1)
  {
  #line 2 SlopeID
    numh =  as.numeric(dat[2])
  # line 3 SlopeType (Constant Thickness or 'Cake shape')
    htyp = as.numeric(dat[3])
  #line 4 Area
    Area <- as.numeric(dat[4])
  # line 5 Number of points
    noPoints <- as.numeric(dat[5])
  # line 6 Thickness of slope
    Depth <- as.numeric(dat[6])
  # line 7 Eta Values
    eta <- as.numeric(na.exclude(as.numeric(strsplit(dat[7]," ")[[1]])))
  # line 8 Xsi Values
    xsi <- as.numeric(na.exclude(as.numeric(strsplit(dat[8]," ")[[1]])))
  # line 9...8+n x of slopepoints
    Rew <- strsplit(dat[9:(8+slopeCount)]," ")
    Rew <- lapply(Rew, function(x){ x <- as.numeric(na.exclude(as.numeric(x)))})
  # line 9+n...8+2n y of slopepoints
    How <- strsplit(dat[(9+slopeCount):(8+2*slopeCount)]," ")
    How <- lapply(How, function(x){ x <- as.numeric(na.exclude(as.numeric(x)))})
  # line 9+2n...8+3n z of slopepoints
    Elev <- strsplit(dat[(9+2*slopeCount):(8+3*slopeCount)]," ")
    Elev <- lapply(Elev, function(x){ x <- as.numeric(na.exclude(as.numeric(x)))})
  # line 9+3n...8+4n width of slopepoints
    Width <- strsplit(dat[(9+3*slopeCount):(8+4*slopeCount)]," ")
    Width <- lapply(Width, function(x){ x <- as.numeric(na.exclude(as.numeric(x)))})

  ## aggregate to slope list
  erg <- list(  "xh" = Rew[[1]], 
                "yh" = How[[1]],
                "zh" = Elev[[1]], 
                "tot.area" = Area,
                "bh" = Width[[1]], 
                "dyy" = Depth,
                "xsi" = unique(sort(xsi)),
                "eta" = unique(sort(eta)),
                "htyp" = htyp ,
                "numh" = numh,
                "ho_bez" = ho_bez,
                "re_bez" = re_bez,
                "z_bez" = z_bez,
                "out.file" = paste("slope", numh, ".geo", sep="")
              )
  
  ### more than one point
  } else {
    #line 2 SlopeID
    numh =  as.numeric(strsplit(dat[2]," ")[[1]])
  # line 3 SlopeType (Constant Thickness or 'Cake shape')
    htyp = as.numeric(strsplit(dat[3]," ")[[1]])
  #line 4 Area
    Area <- as.numeric(na.exclude(as.numeric(strsplit(dat[4]," ")[[1]])))
  # line 5 Number of points
    noPoints <- as.numeric(strsplit(dat[5]," ")[[1]]) +1    ## one point less in file (?)
  # line 6 Thickness of slope
    Depth <- as.numeric(na.exclude(as.numeric(strsplit(dat[6]," ")[[1]])))
  # line 7 Eta Values
    eta <- as.numeric(na.exclude(as.numeric(strsplit(dat[7]," ")[[1]])))
  # line 8 Xsi Values
    xsi <- as.numeric(na.exclude(as.numeric(strsplit(dat[8]," ")[[1]])))
  # line 9...8+n x of slopepoints
    Rew <- strsplit(dat[9:(8+slopeCount)]," ")
    Rew <- lapply(Rew, function(x){ x <- as.numeric(na.exclude(as.numeric(x)))})
  # line 9+n...8+2n y of slopepoints
    How <- strsplit(dat[(9+slopeCount):(8+2*slopeCount)]," ")
    How <- lapply(How, function(x){ x <- as.numeric(na.exclude(as.numeric(x)))})
  # line 9+2n...8+3n z of slopepoints
    Elev <- strsplit(dat[(9+2*slopeCount):(8+3*slopeCount)]," ")
    Elev <- lapply(Elev, function(x){ x <- as.numeric(na.exclude(as.numeric(x)))})
  # line 9+3n...8+4n width of slopepoints
    Width <- strsplit(dat[(9+3*slopeCount):(8+4*slopeCount)]," ")
    Width <- lapply(Width, function(x){ x <- as.numeric(na.exclude(as.numeric(x)))})

  # aggregate to list with one list per hillslope

   for(i in 1:(slopeCount))
   {  erg[[i]] <- list( "xh" = Rew[[i]][1:noPoints[i]], 
                        "yh" = How[[i]][1:noPoints[i]],
                        "zh" = Elev[[i]][1:noPoints[i]], 
                        "tot.area" = Area[i],
                        "bh" = Width[[i]][1:noPoints[i]], 
                        "dyy" = Depth[i],
                        "xsi" = unique(sort(xsi)),
                        "eta" = unique(sort(eta)),
                        "htyp" = htyp[i] ,
                        "numh" = numh[i],
                        "ho_bez" = ho_bez,
                        "re_bez" = re_bez,
                        "z_bez" = z_bez,
                        "out.file" = paste("slope", numh[i], ".geo", sep="")
                        )
   }
   names(erg) <- paste("slope", numh, sep="")
  }
 return(erg)
}

